rectools/metrics/debias.py

```diff
@@ -0,0 +1,118 @@
+        Pseudorandom number generator state to control the down-sampling.
+    """
+
+    iqr_coef: float = attr.ib(default=1.5)
+    random_state: int = attr.ib(default=32)
+
+    @property
+    def keyname(self) -> str:
+        """
+        Compose a config's key name based on the iqr_coef and random_state parameters.
+
+        Returns
+        -------
+        str
+            String representing the key name in the format `"{iqr_coef}_{random_state}"`.
+        """
+        return f"{self.iqr_coef}_{self.random_state}"
+
+
+def make_downsample(interactions: pd.DataFrame, debias_config: DebiasConfig) -> pd.DataFrame:
+    """
+    Downsample the size of interactions, excluding some interactions with popular items.
+
+    Algorithm:
+
+        1. Calculate item popularity distribution from interactions;
+        2. Find first (Q1) and third (Q3) quartiles in items popularity distribution;
+        3. Calculate IQR = Q3 - Q1;
+        4. Calculate maximum value inside by formula: Q3 + iqr_coef * IQR;
+        5. Down-sample for all exceeding items in interactions,
+        randomly keeping the maximum group of users to a size not exceeding
+        maximum value inside
+
+    Parameters
+    ----------
+    interactions : pd.DataFrame
+        Table with previous user-item interactions,
+        with columns `Columns.User`, `Columns.Item`.
+    iqr_coef : float, default 1.5
+        Coefficient for defining as the maximum value inside the border.
+    random_state: float, default 32
+        Pseudorandom number generator state to control the down-sampling.
+
+    Returns
+    -------
+    pd.DataFrame
+        Downsampling interactions.
+    """
+    if len(interactions) == 0:
+        return interactions
+
+    item_popularity = interactions[Columns.Item].value_counts()
+
+    quantiles = item_popularity.quantile(q=[0.25, 0.75])
+    q1, q3 = quantiles.loc[0.25], quantiles.loc[0.75]
+    iqr = q3 - q1
+    max_border = int(q3 + debias_config.iqr_coef * iqr)
+
+    item_outside_max_border = item_popularity[item_popularity > max_border].index
+
+    interactions_result = interactions[~interactions[Columns.Item].isin(item_outside_max_border)]
+    interactions_downsampling = interactions[interactions[Columns.Item].isin(item_outside_max_border)]
+
+    interactions_downsampling = (
+        interactions_downsampling.groupby(Columns.Item, as_index=False)[Columns.User]
+        .agg(lambda users: users.sample(max_border, random_state=debias_config.random_state).tolist())
+        .explode(Columns.User)
+    )
+
+    interactions_result = pd.concat([interactions_result, interactions_downsampling]).sort_values(
+        Columns.User, ignore_index=True
+    )
+
+    interactions_result[Columns.User] = interactions_result[Columns.User].astype(interactions[Columns.User].dtypes)
+    interactions_result[Columns.Item] = interactions_result[Columns.Item].astype(interactions[Columns.Item].dtypes)
+
+    if Columns.Rank in interactions.columns:
```

Comment:
Why can we have rank in interactions?

Reply:
It's the same as interactions, but with an extra `rank` column.
![Снимок экрана 2024-06-07 в 19 12 01](https://github.com/MobileTeleSystems/RecTools/assets/43339130/628b6142-cb5d-4597-abaa-d437a7f89a25)

I added this code, because the separate methods for ranking metrics and classification metrics functions work directly with `merge_reco`. In addition, the structure of the resulting `merge_reco` dataframe doesn't conflict with the debias calculation.

Reply:
Ok, anyway, it seems that all the columns from the original `interactions` table have to be present in the `interactions_result` without any additional merges, no?

Reply:
ok, then I don't understand how to use the debounce mechanism in cases where metrics are counted with `merge_reco` output. For example in these places: [calc_ranking_metrics](https://github.com/MobileTeleSystems/RecTools/blob/main/rectools/metrics/ranking.py#L534) or [calc_from_merged method from Ranking metrics](https://github.com/MobileTeleSystems/RecTools/blob/main/rectools/metrics/ranking.py#L373)

Reply:
Seems we have a misunderstanding here :)

I'll try to rephrase: what will break if you remove this `if` and the code under it?

Reply:
Yes, I see your point now :)
Fixed